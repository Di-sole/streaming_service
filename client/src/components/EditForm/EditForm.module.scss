.container {
    display: flex;

    margin-bottom: 15px;
    padding: 15px;

    border-radius: 8px;
    color: #a5a1b2;

    background-color: #1f1b2e;

    &:hover {
        transform: scale(1.1);
        transition: ease-in-out 0.4s;
    }

    &_visible {
        max-height: 150px;
        opacity: 1;
        transition: max-height 400ms ease, opacity 200ms ease, padding-top 400ms ease;
    }

    &_hidden {
        margin: 0;
        padding: 0;
        max-height: 0;
        opacity: 0;
        overflow: hidden;
        transition: max-height 400ms ease, opacity 200ms ease, padding-top 400ms ease;
    }

    &__form {
        width: auto;
        display: flex;
        justify-content: space-between;
        align-items: center;

        label {
            margin-right: 10px;
            font-size: 14px;
            line-height: 1.5em;
        }

        input {
            box-sizing: border-box;
            padding: 8px 10px;
            height: 20px;

            font-weight: 500;
            font-style: italic;

            border-radius: 40px;
        }

        input:disabled {
            background-color: lightgray;
        }
    }
    &__field {
        margin-right: 15px;

        display: flex;
        align-items: center;

        &_first {
            input {
                width: 100px;
                text-align: center;
            }
        }
    }

    &__editButton,
    &__deleteButton {
        margin-right: 10px;
        padding: 0;
        img {
            margin-bottom: 2px;
        }

        &:hover {
            transform: scale(1.2);
            transition: 0.4s;
        }

        &:active {
            transform: translateY(5px);
        }
    }
    &__deleteButton {
        margin-right: 15px;

        &:hover {
            img {
                filter: invert(67%) sepia(89%) saturate(7492%) hue-rotate(346deg) brightness(84%) contrast(146%);
            }
        }
    }

    &__saveButton {
        transition: 0.4s;

        &:active {
            transform: translateY(5px);
        }
    }
}
