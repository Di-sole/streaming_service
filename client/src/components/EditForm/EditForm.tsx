import Image from 'next/image';
import styles from './EditForm.module.scss';
import { useEffect, useState } from 'react';
import editBtn from '../../../public/icons/edit.svg';
import confirmBtn from '../../../public/icons/confirm.svg';
import deleteBtn from '../../../public/icons/waste_bin.svg';
import BorderedButton from '../UI/buttons/BorderedButton/BorderedButton';
import axios from 'axios';
import { useRouter } from 'next/router';
import { useTranslation } from 'next-i18next';

interface IEditableItem {
    id: number;
    name: string;
    enName: string;
}

interface IEditFormProps {
    item: IEditableItem;
    deletable: boolean;
}

const EditForm = ({ item, deletable }: IEditFormProps) => {
    const { t } = useTranslation('adminPage');
    const [isFieldActive, setIsFieldActive] = useState(false);
    const [isItemDeleted, setIsItemDeleted] = useState(false);

    const [inputValueRus, setInputValueRus] = useState(item.name || '');
    const [inputValueEn, setInputValueEn] = useState(item.enName || '');

    useEffect(() => {
        setInputValueRus(item.name || '');
        setInputValueEn(item.enName || '');
    }, [item]);

    const { asPath } = useRouter();
    const route = asPath.split('/').slice(-1)[0];
    const endpoint = route === 'genres' ? 'namesofgenre' : 'film';

    const handleDeleteItem = async () => {
        try {
            await axios.delete(`http://localhost:6125/film/${item.id}`);
        } catch (e: any) {
            console.log(e.response?.data?.message);
        }
        setIsItemDeleted(true);
    };

    const handleSaveResult = async () => {
        try {
            await axios.patch(`http://localhost:6125/${endpoint}`, {
                id: item.id,
                name: inputValueRus,
                enName: inputValueEn,
            });
        } catch (e: any) {
            console.log(e.response?.data?.message);
        }
        setIsFieldActive(false);
    };
    return (
        <div
            className={[styles.container, isItemDeleted ? styles.container_hidden : styles.container_visible].join(' ')}
        >
            <div className={styles.container__form}>
                {deletable && (
                    <button className={styles.container__deleteButton} onClick={handleDeleteItem}>
                        <Image src={deleteBtn} width={20} height={20} alt="Кнопка удаления" />
                    </button>
                )}
                <div className={[styles.container__field, styles.container__field_first].join(' ')}>
                    <label htmlFor="id">ID</label>
                    <input type="text" id="id" value={item.id} disabled />
                </div>
                <div className={styles.container__field}>
                    <label htmlFor="name-rus">{t('editing.ruName')}</label>
                    <input
                        type="text"
                        id="name-rus"
                        value={inputValueRus}
                        disabled={!isFieldActive}
                        onChange={(e) => setInputValueRus(e.target.value)}
                    />
                </div>
                <div className={styles.container__field}>
                    <label htmlFor="name-en">{t('editing.enName')}</label>
                    <input
                        type="text"
                        id="name-en"
                        value={inputValueEn}
                        disabled={!isFieldActive}
                        onChange={(e) => setInputValueEn(e.target.value)}
                    />
                </div>
                <button className={styles.container__editButton} onClick={() => setIsFieldActive(!isFieldActive)}>
                    <Image
                        src={isFieldActive ? confirmBtn : editBtn}
                        width={20}
                        height={20}
                        alt="Кнопка редактирования и удаления"
                    />
                </button>
                <BorderedButton size="small" className={styles.container__saveButton} onClick={handleSaveResult}>
                    {t('editing.saveButton')}
                </BorderedButton>
            </div>
        </div>
    );
};
export default EditForm;
