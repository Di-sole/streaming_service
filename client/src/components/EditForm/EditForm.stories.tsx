import type { Meta, StoryObj } from '@storybook/react';
import './EditForm.module.scss';
import EditForm from './EditForm';

const meta: Meta<typeof EditForm> = {
    title: 'EditForm/EditForm',
    component: EditForm,
    tags: ['autodocs'],
    parameters: {
        backgrounds: { default: 'dark' },
    },
};

export default meta;

type Story = StoryObj<typeof EditForm>;

export const MovieEditForm: Story = {
    args: {
        item: { id: 1, name: 'Хранители', enName: 'Watchmen' },
        deletable: true,
    },
};
