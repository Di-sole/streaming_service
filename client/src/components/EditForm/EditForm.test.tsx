import { render, screen } from '@testing-library/react';
import EditForm from './EditForm';

jest.mock('next-i18next', () => ({
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        };
    },
}));

jest.mock('next/router', () => ({
    useRouter() {
        return {
            route: '/',
            pathname: '',
            asPath: '/admin/movies',
        };
    },
}));

describe('EDITFORM SNAPSHOTS TESTS', () => {
    test('EditForm should not be changed', () => {
        render(<EditForm item={{ id: 1, name: 'Название', enName: 'Name' }} deletable={true} />);

        const div = screen.getByTestId('editForm');
        expect(div).toMatchSnapshot();
    });
});
