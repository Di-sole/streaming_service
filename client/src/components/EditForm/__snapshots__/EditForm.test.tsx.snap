// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EDITFORM SNAPSHOTS TESTS EditForm should not be changed 1`] = `
<div
  class="container container_visible"
  data-testid="editForm"
>
  <div
    class="container__form"
  >
    <button
      class="container__deleteButton"
    >
      <img
        alt="Кнопка удаления"
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent;"
        width="20"
      />
    </button>
    <div
      class="container__field container__field_first"
    >
      <label
        for="id"
      >
        ID
      </label>
      <input
        disabled=""
        id="id"
        type="text"
        value="1"
      />
    </div>
    <div
      class="container__field"
    >
      <label
        for="name-rus"
      >
        editing.ruName
      </label>
      <input
        disabled=""
        id="name-rus"
        type="text"
        value="Название"
      />
    </div>
    <div
      class="container__field"
    >
      <label
        for="name-en"
      >
        editing.enName
      </label>
      <input
        disabled=""
        id="name-en"
        type="text"
        value="Name"
      />
    </div>
    <button
      class="button button_faded container__editButton"
      type="button"
    >
      <img
        alt="Кнопка редактирования и удаления"
        data-nimg="1"
        decoding="async"
        height="20"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent;"
        width="20"
      />
    </button>
    <button
      class="button button_small container__saveButton"
      type="button"
    >
      editing.saveButton
    </button>
  </div>
</div>
`;
